### ISSUE TITLE
Bun?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/technology-stack#139]
https://bun.sh -> https://github.com/oven-sh/bun

<img width="1325" alt="image" src="https://github.com/user-attachments/assets/7d85e65a-ecb7-4f51-b56c-9756e239dd66">

<img width="1380" alt="image" src="https://github.com/user-attachments/assets/9a3a7552-7215-4ffe-a3a6-f508ec7a0766">

<img width="1381" alt="image" src="https://github.com/user-attachments/assets/5f666bf1-cebd-4160-872b-39e1a49abc54">

Written in `Zig` ref: https://github.com/dwyl/learn-zig/issues/1 

**Note**: We haven't seen a performance bottleneck in our `Node` projects _recently_. 
(mostly because we've not been _using_ `Node` in a perf-critical project for a `while` ...)
But the `Bun` appears to be faster at _everything_ than `Node`. 
So I'm thinking of doing a `SPIKE` just to get it working on a basic project. 💭 
 
I wonder if this is just just a case of "[Faster Horses](https://www.goodreads.com/quotes/15297-if-i-had-asked-people-what-they-wanted-they-would)" ... Or if there is a benefit to this beyond speed?

**Comment by ndrean** [comment]
Check why `Vite` and `Bun` by Evan You (code splitting)

[![Screenshot 2024-07-15 at 15 32 35](https://github.com/user-attachments/assets/39143f01-edce-44e1-b5b8-684586789e89)](https://x.com/youyuxi/status/1701039789741629721)

Bun features Apple's Javascript runtime <https://developer.apple.com/documentation/javascriptcore> instead of V8.

**Comment by nelsonic** [comment]
Yeah, we've never _needed_ to use code splitting. 🐌 
It's a sign that the app is _way_ too bloated. 😬 
Our biggest ever front-end app was `147kb` ... which I felt was silly (slow on mobile with slow 3G ... ⏳)
But it's good to know that `Bun` is not targeting client-side apps. 👌 

`JavaScriptCore` https://github.com/WebKit/WebKit/tree/main/Source/JavaScriptCore is written in `C++`.
So is `V8`: https://nodejs.org/en/learn/getting-started/the-v8-javascript-engine#the-quest-for-performance ... 
There shouldn't be _that_ much of a perf diff ... but all benchmarks say there is a _huge_ diff between `Node` and `Bun` 🤷‍♂️ 
https://www.builder.io/blog/bun-vs-node-js

Again, not using this _yet_. Just opening the issue to capture thoughts. 💭

**Comment by ndrean** [comment]
A disgression, but for example, LiveView ships a lot (!) of (meaningful) Javascript. I believe the code should be split and stripped out from the unused JS. Vite or Rollup (or the new Rust  [RollDown](https://rolldown.rs/) should be used. But this is the work of the team, not ours. Take a look at `Livebook`, JS heavy.  Plenty of React is injected. I don't understand why they didn't go the SolidJS for example, almost 10 times lighter and not virtual DOM. Probably read-to-use libraries, I didn't go into the details.

**Comment by nelsonic** [comment]
Yeah, `Livebook` is _suuuuper_ `JS`-heavy; but it's unlikely to be used on a slow internet connection. 💭 
Even the biggest `JS` Apps are smaller than the average `<img>` ... 
Not saying there's an _excuse_ for bloated JS but there are apps that are complex and need a lot of code. 
_Agree_ that some splitting could help lighten `Livebook` ... 💭 
`bun` might not be the tool for that. 👍

**Comment by ndrean** [comment]
But... they did something amazing with `Kino.JS.Live`. I am not doing any kind of self promotion but take a closer look at the Livebooks I did (any correction/issue is welcome!):

-  [WebRTC-SFU-demo](https://github.com/dwyl/WebRTC-SFU-demo). It showcases the power of message passing, the usage of the underlying WebSocket connection, and the ease to inject a third party JS library, Google's `mediaPipe` library. The result is impressive in this tiny Livebook.

- Another example is with the Livebook [HLS-demo](https://github.com/dwyl/HLS-demo/blob/main/lib/hls-demo.livemd). Again, the power of message passing, the usage of a **raw WebSocket** connection to easily pass binary data, and again the integration of third party JS library. 

I just though 👏👏

**Comment by ndrean** [comment]
`bun` setup with Liveview.

YT link below

[![Screenshot 2024-11-23 at 17 55 22](https://github.com/user-attachments/assets/ab27225f-e888-408e-b66e-97cf7066e32f)](https://www.youtube.com/watch?v=axa9lrzbrSQ)

**Comment by nelsonic** [comment]
Very good overview from Andrew. 👌 
Not clear to me why we would _need_ to go through the effort though ... 🤷‍♂️ 💭

**Comment by ndrean** [comment]
If you are not adding anything, indeed. Otherwise, you can start and stop bloating the JS,  do dynamic imports, in other words, cutting your JS into pieces (all your hooks, web workers code etc), and thus potentially shipping less useless JS.  100k of JS, its not nothing

**Comment by nelsonic** [comment]
definitely not nothing. but feels like a niche problem in a `LiveView` App. 💭 
Still, good to know it's possible and quite straightforward. 👌
